contracts/Daomon721a.sol

```diff
@@ -5,33 +5,269 @@ import "erc721a-upgradeable/contracts/ERC721AUpgradeable.sol";
+     */
+    modifier onlyManager() {
+        require(
+            hasRole(MANAGER_ROLE, msg.sender),
+            "Caller does not have the required role"
+        );
+        _;
+    }
+
+    // =============================================================
+    //                        CONFIGURATION
+    // =============================================================
+
+    /**
+     * Initialize the contract.
+     *
+     * @param _name The name of the token.
+     * @param _symbol The symbol of the token.
+     * @param _tokenUriProvider The tokenURI provider contract which will create the metadata for a token.
+     * @param _unlockPenaltyPercentage The unlock penalty percentage. 0 = no penalty. 10000 = 100% and the minimal penalty is 1 = 0.01 % penalty.
+     */
+    function initialize(
+        string memory _name,
+        string memory _symbol,
+        ITokenURIProvider _tokenUriProvider,
+        uint32 _unlockPenaltyPercentage
+    ) public initializerERC721A initializer {
+        // set roles
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(MANAGER_ROLE, msg.sender);
+
+        // configure the storage
+        tokenUriProvider = _tokenUriProvider;
+        deploymentTimeStamp = block.timestamp;
+        unlockPenaltyPercentage = _unlockPenaltyPercentage;
+
+        // innit all other upgradable contracts
+        __ERC721A_init(_name, _symbol);
         __Ownable_init();
         __AccessControlEnumerable_init();
     }
 
-    /*
-    WRITE METHODS
-    */
+    /**
+     * @notice Sets the optional tokenURI override contract.
+     */
+    function setRenderingContract(
+        ITokenURIProvider _tokenUriProvider
+    ) external onlyManager {
+        require(address(_tokenUriProvider) != address(0), "Invalid address");
+        tokenUriProvider = _tokenUriProvider;
+    }
+
+    /**
+     * @notice Sets the unlock penalty percentage.
+     */
+    function setUnlockPenaltyPercentage(
+        uint32 _unlockPenaltyPercentage
+    ) external onlyManager {
+        require(_unlockPenaltyPercentage >= 0, "Penalty must be >= 0");
+        unlockPenaltyPercentage = _unlockPenaltyPercentage;
+    }
+
+    // =============================================================
+    //                        WRITES
+    // =============================================================
+
+    function toggleLock(uint256 tokenId) public {
+        if (ownerOf(tokenId) != msg.sender) revert NotOwnerOfToken(tokenId);
+
+        // if the token is locked, unlock it and update the total lock time
+        if (isTokenLocked(tokenId)) {
+            lockInfo[tokenId].totalTime = _sanitizeInfo(_getLockInfo(tokenId))
+                .totalTime;
+            lockInfo[tokenId].startTime = 0;
+
+            assert(!isTokenLocked(tokenId));
+        } else {
+            lockInfo[tokenId].startTime = block.timestamp;
+            uint256 penalty = _getPenalty(lockInfo[tokenId].totalTime);
```

Comment:
penalty ist hier beim "lock" implementiert, müsste aber ins "unlock" (also in das "if" nicht in das "else"

Reply:
beim "lock" (als im "else" case) muss eig nur die starttime auf den aktuellen block.timeStamp gesetzt werden

Reply:
I'll double check